src/adapter/src/client.rs

```diff
@@ -396,144 +398,221 @@ impl SessionClient {
-                | ExecuteResponse::DroppedSink
-                | ExecuteResponse::DroppedTable
-                | ExecuteResponse::DroppedView
-                | ExecuteResponse::DroppedMaterializedView
-                | ExecuteResponse::DroppedType
-                | ExecuteResponse::DroppedSecret
-                | ExecuteResponse::DroppedConnection
-                | ExecuteResponse::EmptyQuery
-                | ExecuteResponse::Inserted(_)
-                | ExecuteResponse::StartedTransaction { duplicated: _ }
-                | ExecuteResponse::TransactionExited {
-                    tag: _,
-                    was_implicit: _,
-                }
-                | ExecuteResponse::Updated(_)
-                | ExecuteResponse::AlteredObject(_)
-                | ExecuteResponse::AlteredIndexLogicalCompaction
-                | ExecuteResponse::AlteredSystemConfiguraion
-                | ExecuteResponse::Deallocate { all: _ }
-                | ExecuteResponse::Prepare) => {
-                    results.push(SimpleResult::ok(res));
+                let prep_stmt = match self.get_prepared_statement(&EMPTY_PORTAL).await {
+                    Ok(stmt) => stmt,
+                    Err(err) => {
+                        results.push(SimpleResult::err(err));
+                        continue;
+                    }
+                };
+
+                let param_types = &prep_stmt.desc().param_types;
+                if param_types.len() != raw_params.len() {
+                    let message = format!(
+                        "request supplied {actual} parameters, \
+                         but {statement} requires {expected}",
+                        statement = (&stmt).to_ast_string(),
+                        actual = raw_params.len(),
+                        expected = param_types.len()
+                    );
+                    return Err(AdapterError::Unstructured(anyhow::anyhow!(message)));
                 }
-                ExecuteResponse::SendingRows {
-                    future: rows,
-                    span: _,
-                } => {
-                    let rows = match rows.await {
-                        PeekResponseUnary::Rows(rows) => rows,
-                        PeekResponseUnary::Error(e) => {
-                            results.push(SimpleResult::err(e.to_string()));
-                            continue;
-                        }
-                        PeekResponseUnary::Canceled => {
-                            results
-                                .push(SimpleResult::err("statement canceled due to user request"));
-                            continue;
+
+                let buf = RowArena::new();
+                let mut params = vec![];
+                for (raw_param, mz_typ) in izip!(raw_params.clone(), param_types) {
+                    let pg_typ = mz_pgrepr::Type::from(mz_typ);
+                    let datum = if raw_param.to_uppercase() == "NULL" {
```

Comment:
This looks like it'll lose the ability to distinguish between the string `"null"` and `NULL`. Can we instead use an `Option<&str>` as the parameter value type? That should neatly allow JSON `null` to indicate a `NULL` parameter value.

Reply:
Ah, I had misremembered the ask for numbers to be passed as string as "all values"--fixed